---
title: "Quality Control and Normalization"
output: github_document
---

Created by: Ahmed Mahfouz  
Edited by: Mohammed Charrout, Lieke Michielsen  

```{r, setup, include=FALSE}
knitr::opts_chunk$set(fig.width = 5, fig.height = 5)
options(stringsAsFactors = FALSE)
set.seed(32546)
```

# Overview

In this practical, we will walk through a pipeline to analyze single cell RNA-sequencing (scRNA-seq) data. Starting from a count matrix, we will cover the following steps of the analysis:

1. Quality control 
2. Normalization 
3. Feature selection 

## Datasets

For this tutorial we will use 3 different PBMC datasets from the 10x Genomics website (https://support.10xgenomics.com/single-cell-gene-expression/datasets).

* 1k PBMCs using 10x v2 chemistry
* 1k PBMCs using 10x v3 chemistry
* 1k PBMCs using 10x v3 chemistry in combination with cell surface proteins, but disregarding the protein data and only looking at gene expression.

The datasets are available in this repository. 

Load required packages:

```{r, packages, message=FALSE, warning=FALSE}
library(Seurat)
library(scater)
library(scran)
library(Matrix)
```

# Read the data and create a Seurat object

Here, we use the function `Read10X_h5` of the Seurat package to read in the expression matrices. R stores these matrices as sparse matrix objects, which are essentially memory-efficient tables of values. In this case the values represent the RNA counts in each cell.

```{r, load, warning=FALSE}
v3.1k <- Read10X_h5("pbmc_1k_v3_filtered_feature_bc_matrix.h5")
v2.1k <- Read10X_h5("pbmc_1k_v2_filtered_feature_bc_matrix.h5")
p3.1k <- Read10X_h5("pbmc_1k_protein_v3_filtered_feature_bc_matrix.h5")
# select only gene expression data from the CITE-seq data.
p3.1k <- p3.1k$`Gene Expression`
```

Rather than working directly with matrices, Seurat works with custom objects that wrap around them. 
These Seurat objects also conveniently contain tables of metadata for the cells and features, which avoids the clutter of managing them as separate objects.
As we will see later, normalized expression values are stored in a separate matrix within the Seurat object, which allows us to play around with different normalization strategies without manually keeping a backup of the original values.
In addition to RNA counts, we are able to store additional data types (termed assays) within the Seurat object, such as protein measurements measured by CITE-seq, though we will stick to the default RNA assay here.

First, create Seurat objects for each of the datasets, and then merge into one large Seurat object. We will use the cell metadata to keep track of which dataset the cell originated from.

```{r, create_seurat}
sdata.v2.1k <- CreateSeuratObject(v2.1k, project = "v2.1k")
sdata.v3.1k <- CreateSeuratObject(v3.1k, project = "v3.1k")
sdata.p3.1k <- CreateSeuratObject(p3.1k, project = "p3.1k")
# Merge into one single Seurat object. 
# Prefix cell ids with dataset name (`all.cell.ids`) just in case you have 
# overlapping barcodes between the datasets.
alldata <- merge(sdata.v2.1k, c(sdata.v3.1k, sdata.p3.1k), add.cell.ids=c("v2.1k","v3.1k","p3.1k"))
# Also add in a metadata column that indicates v2 vs v3 chemistry.
chemistry <- rep("v3", ncol(alldata))
chemistry[Idents(alldata) == "v2.1k"] <- "v2"
alldata <- AddMetaData(alldata, chemistry, col.name = "Chemistry")
alldata
```

The metadata of the Seurat object, which itself is a data frame, can be accessed using the slot operator (`@`) like so `alldata@meta.data`.
Alternatively one can call the object with double empty square brackets: `alldata[[]]`.
Another slot to be aware of is `alldata@active.ident`, or alternatively `Idents(alldata)`, which stores a column of the metadata that should be used to identify groups of cells.
The value of the identities are by default chosen to be whatever is passed to the `project` parameter in the `CreateSeuratObject` call, and is stored in the `orig.ident` column of the metadata object.
We are free to change the column that represent the cell identities but for this tutorial (and in the general case) we keep it as is.


Let's check number of cells from each sample using the idents.

```{r, check_samples}
table(Idents(alldata))
```

## 1. Quality control

On object creation, Seurat automatically calculates some QC-stats such as the number of UMIs and features per cell. 
This information is stored in the columns `nCount_RNA` and `nFeature_RNA` of the metadata.

```{r, check_metadata1}
head(alldata@meta.data)
```

Note that the `_RNA` suffix is due to the aforementioned potential to hold multiple assays.
The default assay is named `RNA`, accessible by `alldata[["RNA"]]` or using the assays slot `alldata@assays$RNA`, which is by default set to be the standard active assay (see `alldata@active.assay`).
Effectively this means that any calls that are done on the Seurat object are applied on the `RNA` assay data.

### Calculate mitochondrial proportion

We will manually calculate the proportion of mitochondrial reads and add it to the metadata table.
Mitochondrial genes start with a `MT-` prefix.

```{r, QC1}
percent.mito <- PercentageFeatureSet(alldata, pattern = "^MT-")
alldata <- AddMetaData(alldata, percent.mito, col.name = "percent.mito")
```

#### Calculate ribosomal proportion

In the same manner we will calculate the proportion of the counts that come from ribosomal proteins, identified by the `RPS` and `RPL` prefixes.

```{r QC2}
percent.ribo <- PercentageFeatureSet(alldata, pattern = "^RP[SL]")
alldata <- AddMetaData(alldata, percent.ribo, col.name = "percent.ribo")
```

Now have another look at the metadata table.

```{r check_metadata2}
head(alldata@meta.data)
```

#### Plot QC

Now we can plot some of the QC-features as violin plots. 
Note that Seurat by default will generate a violin plot per identity class.

```{r, plot_qc, fig.width=10, fig.height=8}
VlnPlot(alldata, features = c("nFeature_RNA", "nCount_RNA", "percent.mito", "percent.ribo"), 
        ncol = 2, pt.size = 0.1) + NoLegend()
```

As you can see, the v2 chemistry gives lower gene detection, but higher detection of ribosomal proteins. As the ribosomal proteins are highly expressed they will make up a larger proportion of the transcriptional landscape when fewer of the lowly expressed genes are detected.

We can also plot the different QC-measures as scatter plots.

```{r, plot_scatter1, fig.width = 12, fig.height = 4}
p1 <- FeatureScatter(alldata, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + NoLegend()
p2 <- FeatureScatter(alldata, feature1 = "nFeature_RNA", feature2 = "percent.mito") + NoLegend()
p3 <- FeatureScatter(alldata, feature1="percent.ribo", feature2="nFeature_RNA")
p1 + p2 + p3
```

We can also subset the data to only plot one sample.

```{r, plot_scatter2, fig.height=4, fig.width=6}
FeatureScatter(alldata, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", 
               cells = WhichCells(alldata, expression = orig.ident == "v3.1k") )
```


### Filtering

#### Mitochondrial filtering

We have quite a lot of cells with high proportion of mitochondrial reads. It could be wise to remove those cells, if we have enough cells left after filtering. Another option would be to either remove all mitochondrial reads from the dataset and hope that the remaining genes still have enough biological signal. A third option would be to just regress out the `percent.mito` variable during scaling. 

In this case we have as much as 99.7% mitochondrial reads in some of the cells, so it is quite unlikely that there is much cell type signature left in those.

By eyeballing the plots we can make reasonable decisions on where to draw the cutoff. In this case, the bulk of the cells are below 25% mitochondrial reads and that will be used as a cutoff.

```{r, mito.filt, fig.height=4, fig.width=6}
# Select cells with percent.mito < 25
idx <- which(alldata$percent.mito < 25)
selected <- WhichCells(alldata, cells = idx)
length(selected)
# and subset the object to only keep those cells.
data.filt <- subset(alldata, cells = selected)
# plot violins for new data
VlnPlot(data.filt, features = "percent.mito")
```

As you can see, there is still quite a lot of variation in percent mito, so it will have to be dealt with in the data analysis step.

#### Gene detection filtering

Extremely high number of detected genes could indicate doublets. However, depending on the cell type composition in your sample, you may have cells with higher number of genes (and also higher counts) from one cell type.

In our datasets, we observe a clear difference between the v2 vs v3 10x chemistry with regards to gene detection, so it may not be fair to apply the same cutoffs to all of them. 

Also, in the protein assay data there is a lot of cells with few detected genes giving a bimodal distribution. This type of distribution is not seen in the other 2 datasets. Considering that they are all pbmc datasets it makes sense to regard this distribution as low quality libraries.

Filter the cells with high gene detection (putative doublets) with cutoffs 4100 for v3 chemistry and 2000 for v2.

```{r, gene.filt}
# Start with cells with many genes detected.
high.det.v3 <- WhichCells(data.filt, expression = nFeature_RNA > 4100)
high.det.v2 <- WhichCells(data.filt, expression = nFeature_RNA > 2000 & orig.ident == "v2.1k")
# Remove these cells.
data.filt <- subset(data.filt, cells=setdiff(WhichCells(data.filt),c(high.det.v2,high.det.v3)))
# Check number of cells.
ncol(data.filt)
```

Filter the cells with low gene detection (low quality libraries) with less than 1000 genes for v2 and < 500 for v2.

```{r, gene.filt2}
#start with cells with many genes detected.
low.det.v3 <- WhichCells(data.filt, expression = nFeature_RNA < 1000 & orig.ident != "v2.1k")
low.det.v2 <- WhichCells(data.filt, expression = nFeature_RNA < 500 & orig.ident == "v2.1k")
# remove these cells
data.filt <- subset(data.filt, cells=setdiff(WhichCells(data.filt),c(low.det.v2,low.det.v3)))
# check number of cells
ncol(data.filt)
```

#### Plot QC-stats again

Lets plot the same qc-stats another time.

```{r, vln.plot2, fig.width=10, fig.height=8}
VlnPlot(data.filt, features = c("nFeature_RNA", "nCount_RNA", "percent.mito", "percent.ribo"), 
        ncol = 2, pt.size = 0.1) + NoLegend()
# and check the number of cells per sample before and after filtering
table(Idents(alldata))
table(Idents(data.filt))
```


### Calculate cell-cycle scores 

Seurat has a function for calculating cell cycle scores based on a list of know S-phase and G2/M-phase genes. 

```{r, cc, fig.width=10, warning=FALSE}
data.filt <- CellCycleScoring(
  object = data.filt,
  g2m.features = cc.genes$g2m.genes,
  s.features = cc.genes$s.genes
)
VlnPlot(data.filt, features = c("S.Score","G2M.Score"))
```

In this case it looks like we only have a few cycling cells in the datasets.

## 2. Normalization

To speed things up, we will continue working with the v3.1k dataset only. 
Furthermore, we will switch from working with Seurat to working with the [scater](https://bioconductor.org/packages/release/bioc/vignettes/scater/inst/doc/overview.html) package.
To do so we will convert the Seurat object to a [`SingleCellExperiment`](https://bioconductor.org/packages/devel/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html) (SCE) object, and add some quality controls metrics to filter out low quality cells as before.

```{r, select_dataset}
pbmc.sce <- SingleCellExperiment(assays = list(counts = as.matrix(v3.1k)))
pbmc.sce <- addPerCellQC(pbmc.sce, subsets=list(MT=grepl("^MT-", rownames(pbmc.sce))))
#pbmc.sce <- addPerFeatureQC(pbmc.sce)
```

Similar to Seurat objects, SCE objects hold metadata for the cells and features. 
We can access this data using `colData(pbmc.sce)` and `rowData(pbmc.sce)` respectively.
Take a look at what QC metrics have been calculated for the cells:

```{r, cell_metrics_scater}
head(colData(pbmc.sce))
```
By default we get `sum`, the library size (sum of counts), and `detected`, the number of features with non-zero counts. The column `total` is only relevant if we wish to [subset the genes into distinct groups that are processed seperately](https://bioconductor.org/packages/devel/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#5_Adding_alternative_feature_sets), which we will not do here.
We find similar metrics for just the mitochondrial genes in the `subsets_MT_*` columns as we have specified them explicitly in the `subsets` parameter in the quality control call above.
Note that for subsets we get an additional `*_percent` column indicating the percentage of counts that originate from that gene subset.

We will use these metrics to filter out poor quality cells to avoid negative size factors. 
These steps are very similar to what we have already done on the combined Seurat object but now we perform them on one dataset only using the scater package. 
We can subset SCE objects using the square brackets syntax, as we would normally subset a data frame or matrix in R.

```{r, filter_poor_quality}
pbmc.sce <- pbmc.sce[, pbmc.sce$subsets_MT_percent < 20]
pbmc.sce <- pbmc.sce[, (pbmc.sce$detected > 1000 & pbmc.sce$detected < 4100)]
```

Create a new assay with unnormalized counts for comparison to post-normalization.

```{r, rle_unnormalized, fig.width=6, fig.height=3}
assay(pbmc.sce, "logcounts_raw") <- log2(counts(pbmc.sce) + 1)
plotRLE(pbmc.sce[,1:50], exprs_values = "logcounts_raw", style = "full")
```

Run PCA and save the result in a new object, as we will overwrite the PCA slot later.
Also plot the expression of the B cell marker MS4A1.

```{r, pca_unnormalized, fig.width=9, fig.height=4}
raw.sce <- runPCA(pbmc.sce, exprs_values = "logcounts_raw")
p1 <- scater::plotPCA(raw.sce, colour_by = "total")
p2 <- plotReducedDim(raw.sce, dimred = "PCA", by_exprs_values = "logcounts_raw",
               colour_by = "MS4A1")
p1 + p2
```

### Normalization: Log

In the default normalization method in Seurat, counts for each cell are divided by the total counts for that cell and multiplied by the scale factor 10,000. This is then log transformed.

Here we use the filtered data from the counts slot of the SCE object to create a Seurat object. After normalization, we convert the result back into a SingleCellExperiment object for comparing plots.

```{r, warning=FALSE}
pbmc.seu <- CreateSeuratObject(counts(pbmc.sce), project = "PBMC")
pbmc.seu <- NormalizeData(pbmc.seu)
pbmc.seu.sce <- as.SingleCellExperiment(pbmc.seu)
pbmc.seu.sce <- addPerCellQC(pbmc.seu.sce)
```

Perform PCA and examine the normalization results with `plotRLE` and `plotReducedDim`. 
This time, use `logcounts` as the expression values to plot (or omit the parameter, as `logcounts` is the default value). 
Check some marker genes, for example GNLY (NK cells) or LYZ (monocytes).

```{r, rle_normalized, fig.width=6, fig.height=3}
plotRLE(pbmc.seu.sce[,1:50], style = "full")
```

```{r, pca_normalized, fig.width=9, fig.height=4}
pbmc.seu.sce <- runPCA(pbmc.seu.sce)
p1 <- scater::plotPCA(pbmc.seu.sce, colour_by = "total")
p2 <- plotReducedDim(pbmc.seu.sce, dimred = "PCA", colour_by = "MS4A1")
p1 + p2
```

### Normalization: scran

The normalization procedure in scran is based on the deconvolution method by Lun et al (2016). Counts from many cells are pooled to avoid the drop-out problem. Pool-based size factors are then “deconvolved” into cell-based factors for cell-specific normalization. Clustering cells prior to normalization is not always necessary but it improves normalization accuracy by reducing the number of DE genes between cells in the same cluster.

We will apply this normalization procedure on the unnormalized `pbmc.sce` object from before.

```{r}
qclust <- scran::quickCluster(pbmc.sce)
pbmc.sce <- scran::computeSumFactors(pbmc.sce, clusters = qclust)
summary(sizeFactors(pbmc.sce))
pbmc.sce <- logNormCounts(pbmc.sce)
```

Examine the results and compare to the log-normalized result. Are they different? 

```{r, rle_normalized_scran, fig.width=6, fig.height=3}
plotRLE(pbmc.sce[,1:50], exprs_values = "logcounts", exprs_logged = FALSE, 
        style = "full")
```

```{r, pca_normalized_scran, fig.width=9, fig.height=4}
pbmc.sce <- runPCA(pbmc.sce)
p1 <- scater::plotPCA(pbmc.sce, colour_by = "total")
p2 <- plotReducedDim(pbmc.sce, dimred = "PCA", colour_by = "MS4A1")
p1 + p2
```

## 3. Feature selection

### Feature selection: scran

In the scran method for finding HVGs, a trend is first fitted to the technical variances. In the absence of spike-ins, this is done using the whole data, assuming that the majority of genes are not variably expressed. Then, the biological component of the variance for each endogenous gene is computed by subtracting the fitted value of the trend from the total variance. HVGs are then identified as those genes with the largest biological components. This avoids prioritizing genes that are highly variable due to technical factors such as sampling noise during RNA capture and library preparation. see the [scran vignette](https://bioconductor.org/packages/devel/bioc/vignettes/scran/inst/doc/scran.html#5_variance_modelling) for details.

```{r, fig.width=6, fig.height=4}
dec <- modelGeneVar(pbmc.sce)
dec <- dec[!is.na(dec$FDR),]
top.hvgs <- order(dec$bio, decreasing = TRUE)
head(dec[top.hvgs,])
dec$HVG <- (dec$FDR<0.00001)
hvg_genes <- rownames(dec[dec$FDR < 0.00001, ])
# plot highly variable genes
plot(dec$mean, dec$total, pch=16, cex=0.6, xlab="Mean log-expression",
     ylab="Variance of log-expression")
o <- order(dec$mean)
lines(dec$mean[o], dec$tech[o], col="dodgerblue", lwd=2)
points(dec$mean[dec$HVG], dec$total[dec$HVG], col="red", pch=16)
## save the decomposed variance table and hvg_genes into metadata for safekeeping
metadata(pbmc.sce)$hvg_genes <- hvg_genes
metadata(pbmc.sce)$dec_var <- dec
```

We choose genes that have a biological component that is significantly greater than zero, using a false discovery rate (FDR) of 5%.

```{r, plot_expression, fig.height=4}
plotExpression(pbmc.sce, features = rownames(dec[top.hvgs[1:10],]))
```

### Feature selection: Seurat

The default method in Seurat 3 is variance-stabilizing transformation. A trend is fitted to to predict the variance of each gene as a function of its mean. For each gene, the variance of standardized values is computed across all cells and used to rank the features. By default, 2000 top genes are returned.

```{r, fig.width=8, fig.height=4, warning=FALSE}
pbmc.seu <- FindVariableFeatures(pbmc.seu, selection.method = "vst")
top10 <- head(VariableFeatures(pbmc.seu), 10)
vplot <- VariableFeaturePlot(pbmc.seu)
LabelPoints(plot = vplot, points = top10, repel = TRUE, xnudge = 0, ynudge = 0)
```

Seurat automatically stores the feature metrics in the metadata of the assay.

```{r}
head(pbmc.seu[["RNA"]][[]])
```

How many of the variable genes detected with scran are included in VariableFeatures in Seurat?

```{r}
table(hvg_genes %in% VariableFeatures(pbmc.seu))
```

## Saving the data

We will save the Seurat object for future analysis downstream. 

```{r, export}
saveRDS(pbmc.seu, file = "pbmc3k.rds")
```


## Session info

```{r}
sessionInfo()
```